[DX-1012, TT-11584] Improved endpoint cache docs (#4054)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/middleware/endpoint-cache-tyk-classic.md

---
title: Using the Endpoint Caching middleware with Tyk Classic APIs
date: 2024-01-24
description: "Using the Endpoint Caching middleware with Tyk Classic APIs"
tags: ["Caching", "Cache", "Endpoint Cache", "selective caching", "middleware", "per-endpoint", "Tyk Classic"]
---

The [Endpoint Caching]({{< ref "basic-config-and-security/reduce-latency/caching/advanced-cache" >}}) middleware allows you to perform selective caching for specific endpoints rather than for the entire API, giving you granular control over which paths are cached.

> When working with Tyk Classic APIs the middleware is configured in the Tyk Classic API Definition either manually within the `.json` file or from the API Designer in the Tyk Dashboard.

Author: ```suggestion
When working with Tyk Classic APIs the middleware is configured in the Tyk Classic API Definition. You can do this via the Tyk Dashboard API or in the API Designer.

```

In this example the endpoint caching middleware has been configured to cache all safe requests to two endpoints (`/widget` and `/fish`) with a cache refresh period of 60 seconds.

#### Advanced endpoint cache
For ultimate control over what Tyk caches, you should use the advanced configuration options for the per-endpoint cache. You can separately configure, for each HTTP method for an endpoint:
 - an individual cache refresh (timeout)
 - a list of HTTP response codes that should be cached
 - a pattern match to cache only requests containing specific data in the [request body]({{< ref " basic-config-and-security/reduce-latency/caching/advanced-cache#selective-caching-by-body-value" >}})

To enable the advanced middleware you must add a new `advance_cache_config` object to the `extended_paths` section of your API definition.

In the API-level `cache_options` you must enable caching and ensure that the option to cache all safe requests is disabled. The timeout that you set here will be used as a default  for any endpoints for which you don't want to configure individual timeouts.

The `advance_cache_config` object has the following configuration:
 - `path`: the path to match on

>  - `method`: this method to match on

Reviewer: "the method to match on" ?

 - `timeout`: set to the refresh period for the cache (in seconds)
 - `cache_response_codes`: HTTP responses codes to be cached (for example `200`)
 - `cache_key_regex`: pattern match for selective caching by body value

For example:
```.json  {linenos=true, linenostart=1}
{
    "cache_options": {
        "enable_cache": true,
        "cache_timeout": 60,
        "cache_all_safe_requests": false
    },

    "extended_paths": {
        "advance_cache_config": [
            {
                "disabled": false,
                "method": "POST",
                "path": "/widget",
                "cache_key_regex": "",
                "cache_response_codes": [
                    200
                ],
                "timeout": 10
            }, 
            {
                "disabled": false,
                "method": "GET",
                "path": "/fish",

>                 "cache_key_regex": "shark",

Reviewer: maybe add a more "patern-ish" example like "\"id\":[^,]*" ?

                "cache_response_codes": [
                    200, 300
                ],
                "timeout": 0
            } 
        ]
    }
}
```

In this example the endpoint caching middleware has been configured to cache requests to two endpoints (`/widget` and `/fish`) as follows:

| endpoint | HTTP response codes to cache | cache refresh timeout | body value regex |
|----------|------------------------------|-----------------------|------------------|
| `POST /widget` | 200 | 10 seconds | none |
| `GET /fish` | 200, 300 | 60 seconds (taken from `cache_options`) | `shark` |

## Configuring the middleware in the API Designer
You can use the API Designer in the Tyk Dashboard to configure the endpoint caching middleware for your Tyk Classic API by following these steps.

#### Simple endpoint cache
To enable and configure the simple endpoint cache, follow these instructions:

##### Step 1: Configure the API level caching options
From the **Advanced Options** tab configure the cache as follows:
 - **Enable caching** to enable the cache middleware
 - **Cache timeout** to configure the timeout (in seconds) for cached requests

>  - **Cache only these status codes** is where you list which HTTP status codes should be cached, clicking **Add** after entering a code to add it to the list 

Author: ```suggestion
 - **Cache only these status codes** is a list of HTTP status codes that should be cached, remember to click **Add** after entering each code to add it to the list 
```

##### Step 1: Configure the API level caching options
From the **Advanced Options** tab configure the cache as follows:
 - **Enable caching** to enable the cache middleware
 - **Cache timeout** to configure the default timeout (in seconds) for any endpoints for which you don't want to configure individual timeouts
 - **Cache only these status codes** leave this blank
 - **Cache all safe requests** ensure that this is **not** selected, otherwise the responses from all endpoints for the API will be cached

##### Step 2: Add an endpoint for the path and select the plugin
From the **Endpoint Designer** add an endpoint that matches the path for which you want to cache responses. Select the **Advanced Cache** plugin.

##### Step 3: Configure the Advanced Cache plugin
Set the timeout and HTTP response codes for the endpoint. If you don't need to set a specific timeout for an endpoint you can leave this blank and Tyk will use the cache timeout configured at the API level.

{{< img src="/img/dashboard/endpoint-designer/advanced-cache-config.png" alt="Endpoint cache configuration" >}}

{{< note success >}}
**Note**  

> Body value match caching is not currently exposed in the Dashboard UI, so it must be enabled through either the raw API editor or the Dashboard API. 

Reviewer: configured might sound better than "enabled"
